From 44a1c307077319bab3a8a984718da6a2759e2e02 Mon Sep 17 00:00:00 2001
From: LegendaryFire <tristan.balon@outlook.com>
Date: Wed, 24 Dec 2025 00:54:44 -0800
Subject: [PATCH] hid/multitouch: Prevent mode set on Surface Laptop Studio 2
 touch pad

---
 drivers/hid/hid-multitouch.c | 30 ++++++++++++++++++++++++++++++
 1 file changed, 30 insertions(+)

diff --git a/drivers/hid/hid-multitouch.c b/drivers/hid/hid-multitouch.c
index 0eb4f7d1228e..8f2ed29b4f35 100644
--- a/drivers/hid/hid-multitouch.c
+++ b/drivers/hid/hid-multitouch.c
@@ -82,6 +82,7 @@ MODULE_LICENSE("GPL");
 #define MT_QUIRK_APPLE_TOUCHBAR		BIT(23)
 #define MT_QUIRK_HAS_TYPE_COVER_BACKLIGHT	BIT(24)
 #define MT_QUIRK_HAS_TYPE_COVER_TABLET_MODE_SWITCH	BIT(25)
+#define MT_QUIRK_SKIP_MODESET_ON_HW_OPEN_CLOSE  BIT(26)
 
 #define MT_INPUTMODE_TOUCHSCREEN	0x02
 #define MT_INPUTMODE_TOUCHPAD		0x03
@@ -242,6 +243,7 @@ static void mt_post_parse(struct mt_device *td, struct mt_application *app);
 #define MT_CLS_SMART_TECH			0x0113
 #define MT_CLS_APPLE_TOUCHBAR			0x0114
 #define MT_CLS_WIN_8_MS_SURFACE_TYPE_COVER	0x0115
+#define MT_CLS_SURFACE_TOUCHPAD                0x0116
 #define MT_CLS_SIS				0x0457
 
 #define MT_DEFAULT_MAXCONTACT	10
@@ -449,6 +451,10 @@ static const struct mt_class mt_classes[] = {
 			MT_QUIRK_WIN8_PTP_BUTTONS,
 		.export_all_inputs = true
 	},
+	{ .name = MT_CLS_SURFACE_TOUCHPAD,
+		.quirks = MT_QUIRK_ALWAYS_VALID |
+			MT_QUIRK_SKIP_MODESET_ON_HW_OPEN_CLOSE
+	},
 	{ }
 };
 
@@ -2180,11 +2186,30 @@ static void mt_remove(struct hid_device *hdev)
 
 static void mt_on_hid_hw_open(struct hid_device *hdev)
 {
+	/*
+	 * Some devices (e.g. Surface Laptop Studio 2 touchpad) can get stuck
+	 * non-functional if we change touchpad reporting modes from the HID
+	 * open/close hooks. Avoid mode switching on hw_open/hw_close for
+	 * those devices.
+	 */
+	struct mt_device *td = hid_get_drvdata(hdev);
+	if (td && td->mtclass.quirks & MT_QUIRK_SKIP_MODESET_ON_HW_OPEN_CLOSE)
+		return;
 	mt_set_modes(hdev, HID_LATENCY_NORMAL, TOUCHPAD_REPORT_ALL);
 }
 
 static void mt_on_hid_hw_close(struct hid_device *hdev)
 {
+	/*
+	 * Some devices (e.g. Surface Laptop Studio 2 touchpad) can get stuck
+	 * non-functional if we change touchpad reporting modes from the HID
+	 * open/close hooks. Avoid mode switching on hw_open/hw_close for
+	 * those devices.
+	 */
+	struct mt_device *td = hid_get_drvdata(hdev);
+	if (td && td->mtclass.quirks & MT_QUIRK_SKIP_MODESET_ON_HW_OPEN_CLOSE)
+		return;
+
 	mt_set_modes(hdev, HID_LATENCY_HIGH, TOUCHPAD_REPORT_NONE);
 }
 
@@ -2621,6 +2646,11 @@ static const struct hid_device_id mt_devices[] = {
 		HID_DEVICE(HID_BUS_ANY, HID_GROUP_ANY,
 			USB_VENDOR_ID_MICROSOFT, 0x09c0) },
 
+	/* Microsoft Surface touch pad */
+	{ .driver_data = MT_CLS_SURFACE_TOUCHPAD,
+		HID_DEVICE(HID_BUS_ANY, HID_GROUP_ANY,
+			USB_VENDOR_ID_MICROSOFT, 0x0C46) },
+
 	/* Google MT devices */
 	{ .driver_data = MT_CLS_GOOGLE,
 		HID_DEVICE(HID_BUS_ANY, HID_GROUP_ANY, USB_VENDOR_ID_GOOGLE,
-- 
2.52.0

