From 9bb4a6a8a1aef64e4fcaa3527108b8e6da35e845 Mon Sep 17 00:00:00 2001
From: Maximilian Luz <luzmaximilian@gmail.com>
Date: Sat, 28 Sep 2019 17:59:13 +0200
Subject: [PATCH 07/13] sdcard-reader

---
 drivers/usb/core/hub.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/drivers/usb/core/hub.c b/drivers/usb/core/hub.c
index d5fbd36cf462..606e63691d50 100644
--- a/drivers/usb/core/hub.c
+++ b/drivers/usb/core/hub.c
@@ -4171,7 +4171,8 @@ void usb_enable_lpm(struct usb_device *udev)
 	if (!udev || !udev->parent ||
 			udev->speed < USB_SPEED_SUPER ||
 			!udev->lpm_capable ||
-			udev->state < USB_STATE_DEFAULT)
+			udev->state < USB_STATE_DEFAULT ||
+			!udev->bos || !udev->bos->ss_cap)
 		return;
 
 	udev->lpm_disable_count--;
-- 
2.24.0

