From eb1ef50f9b1169e3395fbccbdf485aa0e96f19dd Mon Sep 17 00:00:00 2001
From: Maximilian Luz <luzmaximilian@gmail.com>
Date: Sat, 28 Sep 2019 17:59:13 +0200
Subject: [PATCH 07/13] sdcard-reader

---
 drivers/usb/core/hub.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/drivers/usb/core/hub.c b/drivers/usb/core/hub.c
index 8018f813972e..4f7779c53ab2 100644
--- a/drivers/usb/core/hub.c
+++ b/drivers/usb/core/hub.c
@@ -4164,7 +4164,8 @@ void usb_enable_lpm(struct usb_device *udev)
 	if (!udev || !udev->parent ||
 			udev->speed < USB_SPEED_SUPER ||
 			!udev->lpm_capable ||
-			udev->state < USB_STATE_DEFAULT)
+			udev->state < USB_STATE_DEFAULT ||
+			!udev->bos || !udev->bos->ss_cap)
 		return;
 
 	udev->lpm_disable_count--;
-- 
2.23.0

