From 83e9351691f762b82d92a9939ed84798eda3086f Mon Sep 17 00:00:00 2001
From: Dorian Stoll <dorian.stoll@tmsp.io>
Date: Fri, 27 Sep 2019 15:34:51 +0200
Subject: [PATCH] update meson

---
 meson.build | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/meson.build b/meson.build
index 28b46ed..9470e95 100644
--- a/meson.build
+++ b/meson.build
@@ -306,6 +306,12 @@ data_files = [
 	'data/one-by-wacom-s-p.tablet',
 	'data/one-by-wacom-s-p2.tablet',
 	'data/serial-wacf004.tablet',
+	'data/surface-book.tablet',
+	'data/surface-book2-13.tablet',
+	'data/surface-book2-15.tablet',
+	'data/surface-pro4.tablet',
+	'data/surface-pro5.tablet',
+	'data/surface-pro6.tablet',
 	'data/xp-pen-star03.tablet',
 ]
 
-- 
2.23.0

