From 5fc0582727cb64230d1c2ac02001263f92a9b82c Mon Sep 17 00:00:00 2001
From: Dorian Stoll <dorian.stoll@tmsp.io>
Date: Sun, 14 May 2023 09:21:43 +0200
Subject: [PATCH] Pull in iptsd as a weak dependency

Signed-off-by: Dorian Stoll <dorian.stoll@tmsp.io>
---
 redhat/kernel.spec.template | 1 +
 1 file changed, 1 insertion(+)

diff --git a/redhat/kernel.spec.template b/redhat/kernel.spec.template
index 51f43b21b018..d61276b49d95 100644
--- a/redhat/kernel.spec.template
+++ b/redhat/kernel.spec.template
@@ -938,6 +938,7 @@ Requires(pre): %{kernel_prereq}\
 Requires(pre): %{initrd_prereq}\
 Requires(pre): ((linux-firmware >= 20150904-56.git6ebf5d57) if linux-firmware)\
 Recommends: linux-firmware\
+Recommends: iptsd\
 Requires(preun): systemd >= 200\
 Conflicts: xfsprogs < 4.3.0-1\
 Conflicts: xorg-x11-drv-vmmouse < 13.0.99\
-- 
2.40.1

