From 1cad947df5f2ec874863e42cce595a76dad64cc2 Mon Sep 17 00:00:00 2001
From: Dorian Stoll <dorian.stoll@tmsp.io>
Date: Sat, 22 Jul 2023 10:33:03 +0200
Subject: [PATCH] Pull in iptsd as a weak dependency

Signed-off-by: Dorian Stoll <dorian.stoll@tmsp.io>
---
 redhat/kernel.spec.template | 1 +
 1 file changed, 1 insertion(+)

diff --git a/redhat/kernel.spec.template b/redhat/kernel.spec.template
index d3dafe56aa64..24823cd8b784 100644
--- a/redhat/kernel.spec.template
+++ b/redhat/kernel.spec.template
@@ -1052,6 +1052,7 @@ Requires(pre): %{kernel_prereq}\
 Requires(pre): %{initrd_prereq}\
 Requires(pre): ((linux-firmware >= 20150904-56.git6ebf5d57) if linux-firmware)\
 Recommends: linux-firmware\
+Recommends: iptsd\
 Requires(preun): systemd >= 200\
 Conflicts: xfsprogs < 4.3.0-1\
 Conflicts: xorg-x11-drv-vmmouse < 13.0.99\
-- 
2.41.0

