From ee2289e8c6a72b5e778fb0f16bc804ca290cf868 Mon Sep 17 00:00:00 2001
From: Dorian Stoll <dorian.stoll@tmsp.io>
Date: Sun, 14 May 2023 18:16:38 +0200
Subject: [PATCH] Let kernel-surface conflict with older versions of the
 package

Signed-off-by: Dorian Stoll <dorian.stoll@tmsp.io>
---
 redhat/kernel.spec.template | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/redhat/kernel.spec.template b/redhat/kernel.spec.template
index d952180f91ef..e59589949020 100644
--- a/redhat/kernel.spec.template
+++ b/redhat/kernel.spec.template
@@ -599,6 +599,10 @@ Requires: kernel-modules-uname-r = %{KVERREL}
 Requires: kernel-modules-core-uname-r = %{KVERREL}
 Provides: installonlypkg(kernel)
 Provides: kernel = %{KVERREL}
+
+Conflicts: kernel-surface < 6.3.6-1
+Obsoletes: kernel-surface < 6.3.6-1
+
 %endif
 
 
@@ -1230,6 +1234,8 @@ Provides: kernel-devel-%{_target_cpu} = %{specrpmversion}-%{release}%{?1:+%{1}}\
 Provides: kernel-devel-uname-r = %{KVERREL}%{?1:+%{1}}\
 Provides: kernel-devel = %{KVERREL}%{?1:+%{1}}\
 Provides: installonlypkg(kernel)\
+Conflicts: kernel-surface-devel < 6.3.6-1\
+Obsoletes: kernel-surface-devel < 6.3.6-1\
 AutoReqProv: no\
 Requires(pre): findutils\
 Requires: findutils\
-- 
2.40.1

