From cb7cebc53abc4152d3db1e3e53b069792dd3171e Mon Sep 17 00:00:00 2001
From: Dorian Stoll <dorian.stoll@tmsp.io>
Date: Sat, 22 Jul 2023 10:40:32 +0200
Subject: [PATCH 3/5] Let kernel-surface conflict with older versions of the
 package

Signed-off-by: Dorian Stoll <dorian.stoll@tmsp.io>
---
 redhat/kernel.spec.template | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/redhat/kernel.spec.template b/redhat/kernel.spec.template
index ca6517bdc945..2be57434622c 100644
--- a/redhat/kernel.spec.template
+++ b/redhat/kernel.spec.template
@@ -724,6 +724,10 @@ Requires: kernel-modules-core-uname-r = %{KVERREL}
 Requires: ((kernel-modules-extra-uname-r = %{KVERREL}) if kernel-modules-extra-matched)
 Provides: installonlypkg(kernel)
 Provides: kernel = %{KVERREL}
+
+Conflicts: kernel-surface < 6.3.6-1
+Obsoletes: kernel-surface < 6.3.6-1
+
 %endif
 
 
@@ -1443,6 +1447,8 @@ Provides: kernel-devel-%{_target_cpu} = %{specrpmversion}-%{release}%{uname_suff
 Provides: kernel-devel-uname-r = %{KVERREL}%{uname_suffix %{?1:+%{1}}}\
 Provides: kernel-devel = %{KVERREL}%{?1:+%{1}}\
 Provides: installonlypkg(kernel)\
+Conflicts: %{package_name}-devel < 6.3.7-2\
+Obsoletes: %{package_name}-devel < 6.3.7-2\
 AutoReqProv: no\
 Requires(pre): findutils\
 Requires: findutils\
@@ -1472,6 +1478,8 @@ Summary: Meta package to install matching core and devel packages for a given %{
 Requires: %{package_name}%{?1:-%{1}}-devel = %{specrpmversion}-%{release}\
 Requires: %{package_name}%{?1:-%{1}}-core = %{specrpmversion}-%{release}\
 Provides: kernel-devel-matched = %{specrpmversion}-%{release}\
+Conflicts: %{package_name}-devel-matched < 6.3.7-2\
+Obsoletes: %{package_name}-devel-matched < 6.3.7-2\
 %description %{?1:%{1}-}devel-matched\
 This meta package is used to install matching core and devel packages for a given %{?2:%{2} }kernel.\
 %{nil}
@@ -1520,6 +1528,8 @@ Provides: kernel%{?1:-%{1}}-modules-extra-uname-r = %{KVERREL}%{uname_suffix %{?
 Requires: kernel-uname-r = %{KVERREL}%{uname_suffix %{?1:+%{1}}}\
 Requires: kernel%{?1:-%{1}}-modules-uname-r = %{KVERREL}%{uname_suffix %{?1:+%{1}}}\
 Requires: kernel%{?1:-%{1}}-modules-core-uname-r = %{KVERREL}%{uname_suffix %{?1:+%{1}}}\
+Conflicts: %{package_name}-modules-extra < 6.3.7-2\
+Obsoletes: %{package_name}-modules-extra < 6.3.7-2\
 %if %{-m:1}%{!-m:0}\
 Requires: kernel-modules-extra-uname-r = %{KVERREL}%{uname_variant %{?1:+%{1}}}\
 %endif\
@@ -1543,6 +1553,8 @@ Provides: installonlypkg(kernel-module)\
 Provides: kernel%{?1:-%{1}}-modules-uname-r = %{KVERREL}%{uname_suffix %{?1:+%{1}}}\
 Requires: kernel-uname-r = %{KVERREL}%{uname_suffix %{?1:+%{1}}}\
 Requires: kernel%{?1:-%{1}}-modules-core-uname-r = %{KVERREL}%{uname_suffix %{?1:+%{1}}}\
+Conflicts: %{package_name}-modules < 6.3.7-2\
+Obsoletes: %{package_name}-modules < 6.3.7-2\
 %if %{-m:1}%{!-m:0}\
 Requires: kernel-modules-uname-r = %{KVERREL}%{uname_variant %{?1:+%{1}}}\
 %endif\
@@ -1565,6 +1577,8 @@ Provides: kernel-modules-core = %{specrpmversion}-%{release}%{uname_suffix %{?1:
 Provides: installonlypkg(kernel-module)\
 Provides: kernel%{?1:-%{1}}-modules-core-uname-r = %{KVERREL}%{uname_suffix %{?1:+%{1}}}\
 Requires: kernel-uname-r = %{KVERREL}%{uname_suffix %{?1:+%{1}}}\
+Conflicts: %{package_name}-modules-core < 6.3.7-2\
+Obsoletes: %{package_name}-modules-core < 6.3.7-2\
 %if %{-m:1}%{!-m:0}\
 Requires: kernel-modules-core-uname-r = %{KVERREL}%{uname_variant %{?1:+%{1}}}\
 %endif\
@@ -1588,6 +1602,8 @@ Requires: ((kernel-%{1}-modules-extra-uname-r = %{KVERREL}%{uname_suffix %{1}})
 %if "%{1}" == "rt" || "%{1}" == "rt-debug" || "%{1}" == "rt-64k" || "%{1}" == "rt-64k-debug"\
 Requires: realtime-setup\
 %endif\
+Conflicts: %{package_name} < 6.3.7-2\
+Obsoletes: %{package_name} < 6.3.7-2\
 Provides: installonlypkg(kernel)\
 Provides: kernel = %{KVERREL}+%{1}\
 %description %{1}\
@@ -1605,6 +1621,8 @@ Summary: %{variant_summary}\
 Provides: kernel-%{?1:%{1}-}core-uname-r = %{KVERREL}%{uname_suffix %{?1:+%{1}}}\
 Provides: installonlypkg(kernel)\
 Provides: kernel-%{?1:%{1}-}core = %{KVERREL}%{?1:+%{1}}\
+Conflicts: %{package_name}-core < 6.3.7-2\
+Obsoletes: %{package_name}-core < 6.3.7-2\
 %if %{-m:1}%{!-m:0}\
 Requires: kernel-core-uname-r = %{KVERREL}%{uname_variant %{?1:+%{1}}}\
 Requires: kernel-%{?1:%{1}-}-modules-core-uname-r = %{KVERREL}%{uname_variant %{?1:+%{1}}}\
-- 
2.50.0

