From 1fed6b82af53c4949cb64a53ba4ecc080a764e25 Mon Sep 17 00:00:00 2001
From: Dorian Stoll <dorian.stoll@tmsp.io>
Date: Tue, 30 Sep 2025 20:10:38 +0200
Subject: [PATCH 3/6] Let kernel-surface conflict with older versions of the
 package

Signed-off-by: Dorian Stoll <dorian.stoll@tmsp.io>
---
 redhat/kernel.spec.template | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/redhat/kernel.spec.template b/redhat/kernel.spec.template
index d92ef1a8d86a..c5748edf9301 100644
--- a/redhat/kernel.spec.template
+++ b/redhat/kernel.spec.template
@@ -785,6 +785,10 @@ Requires: ((%{name}-modules-extra-uname-r = %{KVERREL}) if %{name}-modules-extra
 Suggests: %{name}-core = %{specversion}-%{release}
 Provides: installonlypkg(kernel)
 Provides: kernel = %{KVERREL}
+
+Conflicts: kernel-surface < 6.3.6-1
+Obsoletes: kernel-surface < 6.3.6-1
+
 %endif
 
 
@@ -1619,6 +1623,8 @@ Provides: kernel-devel-uname-r = %{KVERREL}%{uname_suffix %{?1}}\
 Provides: %{name}-devel-uname-r = %{KVERREL}%{uname_suffix %{?1}}\
 Provides: kernel-devel = %{KVERREL}%{uname_suffix %{?1}}\
 Provides: installonlypkg(kernel)\
+Conflicts: %{package_name}-devel < 6.3.7-2\
+Obsoletes: %{package_name}-devel < 6.3.7-2\
 AutoReqProv: no\
 Requires(pre): findutils\
 Requires: findutils\
@@ -1648,6 +1654,8 @@ Summary: Meta package to install matching core and devel packages for a given %{
 Requires: %{name}%{?1:-%{1}}-devel = %{specrpmversion}-%{release}\
 Requires: %{name}%{?1:-%{1}}-core = %{specrpmversion}-%{release}\
 Provides: kernel-devel-matched = %{specrpmversion}-%{release}\
+Conflicts: %{package_name}-devel-matched < 6.3.7-2\
+Obsoletes: %{package_name}-devel-matched < 6.3.7-2\
 %description %{?1:%{1}-}devel-matched\
 This meta package is used to install matching core and devel packages for a given %{?2:%{2} }kernel.\
 %{nil}
@@ -1710,6 +1718,8 @@ Provides: %{name}%{?1:-%{1}}-modules-extra-uname-r = %{KVERREL}%{uname_suffix %{
 Requires: %{name}-uname-r = %{KVERREL}%{uname_suffix %{?1}}\
 Requires: %{name}%{?1:-%{1}}-modules-uname-r = %{KVERREL}%{uname_suffix %{?1}}\
 Requires: %{name}%{?1:-%{1}}-modules-core-uname-r = %{KVERREL}%{uname_suffix %{?1}}\
+Conflicts: %{package_name}-modules-extra < 6.3.7-2\
+Obsoletes: %{package_name}-modules-extra < 6.3.7-2\
 %if %{-m:1}%{!-m:0}\
 Requires: %{name}-modules-extra-uname-r = %{KVERREL}%{uname_variant %{?1}}\
 %endif\
@@ -1743,6 +1753,8 @@ Provides: kernel%{?1:-%{1}}-modules-uname-r = %{KVERREL}%{uname_suffix %{?1}}\
 Provides: %{name}%{?1:-%{1}}-modules-uname-r = %{KVERREL}%{uname_suffix %{?1}}\
 Requires: %{name}-uname-r = %{KVERREL}%{uname_suffix %{?1}}\
 Requires: %{name}%{?1:-%{1}}-modules-core-uname-r = %{KVERREL}%{uname_suffix %{?1}}\
+Conflicts: %{package_name}-modules < 6.3.7-2\
+Obsoletes: %{package_name}-modules < 6.3.7-2\
 %if %{-m:1}%{!-m:0}\
 Requires: %{name}-modules-uname-r = %{KVERREL}%{uname_variant %{?1}}\
 %endif\
@@ -1769,6 +1781,8 @@ Provides: installonlypkg(kernel-module)\
 Provides: kernel%{?1:-%{1}}-modules-core-uname-r = %{KVERREL}%{uname_suffix %{?1}}\
 Provides: %{name}%{?1:-%{1}}-modules-core-uname-r = %{KVERREL}%{uname_suffix %{?1}}\
 Requires: %{name}-uname-r = %{KVERREL}%{uname_suffix %{?1}}\
+Conflicts: %{package_name}-modules-core < 6.3.7-2\
+Obsoletes: %{package_name}-modules-core < 6.3.7-2\
 %if %{-m:1}%{!-m:0}\
 Requires: %{name}-modules-core-uname-r = %{KVERREL}%{uname_variant %{?1}}\
 %endif\
@@ -1794,6 +1808,8 @@ Suggests: %{name}-%{1}-core = %{specversion}-%{release}\
 %if "%{1}" == "rt" || "%{1}" == "rt-debug" || "%{1}" == "rt-64k" || "%{1}" == "rt-64k-debug"\
 Requires: realtime-setup\
 %endif\
+Conflicts: %{package_name} < 6.3.7-2\
+Obsoletes: %{package_name} < 6.3.7-2\
 Provides: installonlypkg(kernel)\
 Provides: kernel = %{KVERREL}+%{1}\
 %description %{1}\
@@ -1821,6 +1837,8 @@ Provides: kernel-%{?1:%{1}-}core-uname-r = %{KVERREL}%{uname_suffix %{?1}}\
 Provides: %{name}-%{?1:%{1}-}core-uname-r = %{KVERREL}%{uname_suffix %{?1}}\
 Provides: installonlypkg(kernel)\
 Provides: kernel-%{?1:%{1}-}core = %{KVERREL}%{uname_suffix %{?1}}\
+Conflicts: %{package_name}-core < 6.3.7-2\
+Obsoletes: %{package_name}-core < 6.3.7-2\
 %if %{-m:1}%{!-m:0}\
 Requires: %{name}-core-uname-r = %{KVERREL}%{uname_variant %{?1}}\
 Requires: %{name}-%{?1:%{1}-}-modules-core-uname-r = %{KVERREL}%{uname_variant %{?1}}\
-- 
2.53.0

