From 1a1b787441b355ea9bbcaf335d3c3ae5ca44534f Mon Sep 17 00:00:00 2001
From: "Justin M. Forbes" <jforbes@fedoraproject.org>
Date: Wed, 12 Jul 2023 07:34:54 -0500
Subject: [PATCH] Add hid-uclogic-test to mod-internal.list

Signed-off-by: Justin M. Forbes <jforbes@fedoraproject.org>
---
 redhat/scripts/mod/mod-internal.list | 1 +
 1 file changed, 1 insertion(+)

diff --git a/redhat/scripts/mod/mod-internal.list b/redhat/scripts/mod/mod-internal.list
index b0368901ff30..3bebd2dfc05c 100644
--- a/redhat/scripts/mod/mod-internal.list
+++ b/redhat/scripts/mod/mod-internal.list
@@ -26,6 +26,7 @@ fat_test
 fortify_kunit
 gss_krb5_test
 hashtable_test
+hid-uclogic-test
 iio-test-format
 iio-test-rescale
 is_signed_type_kunit
-- 
2.41.0

