From 62bc6c700904071cd13e3baa8117e94f7e6a571d Mon Sep 17 00:00:00 2001
From: Maximilian Luz <luzmaximilian@gmail.com>
Date: Sun, 20 Apr 2025 03:02:20 +0200
Subject: [PATCH] kbuild: Copy config to target directory

---
 scripts/package/install-extmod-build | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/scripts/package/install-extmod-build b/scripts/package/install-extmod-build
index b96538787f3d..718bd64d060e 100755
--- a/scripts/package/install-extmod-build
+++ b/scripts/package/install-extmod-build
@@ -68,4 +68,7 @@ if [ "${CC}" != "${HOSTCC}" ]; then
 	rm -f "${destdir}/scripts/Kbuild"
 fi
 
+# copy .config manually to be where it's expected to be
+cp "${KCONFIG_CONFIG}" "${destdir}/.config"
+
 find "${destdir}" \( -name '.*.cmd' -o -name '*.o' \) -delete
-- 
2.51.0

